docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+    public static void StoreUnsafe<T>(this Vector128<T> source, ref T destination) where T : struct;
+    public static void StoreUnsafe<T>(this Vector128<T> source, ref T destination, nuint elementOffset) where T : struct;
+}
+```
+
+For the reasons described for loading, we recommend using the overload that takes managed reference and element offset:
+
+```cs
+public static void StoreUnsafe<T>(this Vector128<T> source, ref T destination, nuint elementOffset) where T : struct;
+```
+
+**Note**: when loading values from one buffer and storing them into another, we need to consider whether they overlap or not. [MemoryExtensions.Overlap](https://learn.microsoft.com/dotnet/api/system.memoryextensions.overlaps#system-memoryextensions-overlaps-1(system-readonlyspan((-0))-system-readonlyspan((-0)))) is an API for doing that.
+
+### Casting
+
+As mentioned before, `Vector128<T>` and `Vector256<T>` are constrained to a specific set of primitive types. Currently, `char` is not one of them, but it does not mean that we can't implement vectorized text operations with the new APIs. For primitive types of the same size (and value types that don't contain references), casting is the solution.
+
+[Unsafe.As<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.unsafe.as#system-runtime-compilerservices-unsafe-as-2(-0@)) can be used to get a reference to supported type:
+
+```cs
+void CastingReferences(Span<char> buffer)
+{
+    ref char charSearchSpace = ref MemoryMarshal.GetReference(buffer);
+    ref short searchSpace = ref Unsafe.As<char, short>(ref charSearchSpace);
+    // from now on we can use Vector128<short> or Vector256<short>
+}
+```
+
+Or [MemoryMarshal.Cast<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.memorymarshal.cast#system-runtime-interopservices-memorymarshal-cast-2(system-readonlyspan((-0)))), which casts a span of one primitive type to a span of another primitive type:
+
+```cs
+void CastingSpans(Span<char> chars)
+{
+    Span<short> shorts = MemoryMarshal.Cast<char, short>(chars);
+}
+```
+
+It's also possible to get managed references from unmanaged pointers:
+
+```cs
+void PointerToReference(char* pUtf16Buffer, byte* pAsciiBuffer)
+{
+    // of the same type:
+    ref byte asciiBuffer = ref *pAsciiBuffer;
+    // of different types:
+    ref ushort utf16Buffer = ref *(ushort*)pUtf16Buffer;
+}
+```
+
+It's only safe to convert a managed reference to a pointer if it's known that the reference is already pinned. If it's not, the moment after you get the pointer it could be invalid.
+
+## Mindset
+
+Vectorizing real-world algorithms seems complex at the beginning. And what do software engineers do with complex problems? We break them down into sub-problems until these become simple enough to be solved directly.
+
+Let's implement a vectorized method for checking whether a given byte buffer consists only from valid ASCII characters to see how similar problems can be solved.
```

Comment:
I think it would be better to have the example of vectorization earlier in the doc.

Basically:
1. Brief introduction to what vectors are and what .NET provides at a high level
2. An example of taking a scalar algorithm and vectorizing it, showing perf wins and touching briefly on what we did
3. A more in depth coverage of the APIs we provide, best practices, etc
4. A more in depth coverage of what not to do and why
5. Optionally some coverage of advanced topics/scenarios, such as handling trailing elements or alignment efficiently

Reply:
This different ordering would walk users through a very natural process introducing key terms, giving an example of what can be done, and then breaking it apart and going into more detail.

As we have it now, there is a lot of detail up front before we ever get into showing a user the "simple example" of taking a scalar algorithm and vectorizing it.

Reply:
> I think it would be better to have the example of vectorization earlier in the doc.

In a way we already have that, as in the "Loops" paragraph I provide a working `Sum` and `Contains` method implementations with comments.

I've spent a lot of time thinking about your recommendation and I believe it's a good one, but I would prefer to keep the existing order.

My main motivation is to prevent the readers from running into various traps, so I really want them to read the whole document. I am hoping that with the current structure, our target readers are going to learn something interesting in every paragraph, which is going to keep them focused and reach the end without missing anything.

With your proposal, I am afraid that some of the readers could skip some parts of it and run into issues later.